File: documentModel.py
from graphicalobject import GraphicalObjectListener, GraphicalObject
from documentModelListener import DocumentModelListener
from point import Point


class ConcreteGraphicalObjectListener(GraphicalObjectListener, DocumentModel):

    def graphicalObjectSelectionChanged(go: GraphicalObject):
        index = DocumentModel._selectedObjects.index(go)
        if go.isSelected():
            if index == -1:
                DocumentModel._selectedObjects.append(go)
            else:
                return
        else:
            if index != -1:
                DocumentModel._selectedObjects.remove(go)
            else:
                return
        self.notifyListeners()

    def graphicalObjectChanged(go: GraphicalObject):
        self.notifyListeners()

class DocumentModel:

    SELECTION_PROXIMITY = 10

    #Kolekcija svih grafičkih objekata.
    _objects = None
    #Proxy oko kolekcije grafičkih objekata.
    _roObjects = None
    #Kolekcija prijavljenih promatrača
    _listeners = None
    #Kolekcija selektiranih objekata.
    _selectedObjects = None
    #Proxy oko kolekcije selektiranih objekata.
    _roSelectedObjects = None
    
    #Promatrač koji će biti registriran nad svim objektima crteža.
    _goListener = ConcreteGraphicalObjectListener()

    def __init__(self):
        self._objects = []
        self._roObjects = [self._objects]
        self._selectedObjects = []
        self._roSelectedObjects = [self._selectedObjects]
        self._listeners = []

    #Brisanje svih objekata iz modela.
    def clear(self):
        for go in self._objects:
            go.removeGraphicalObjectListener(self._goListener)
        
        self._objects.clear()
        self._selectedObjects.clear()
        self.notifyListeners()

    #Dodavanje objekta u dokument.
    def addGraphicalObject(self, go: GraphicalObject):
        if go.isSelected():
            self._selectedObjects.append(go)
        
        self._objects.append(go)
        go.addGraphicalObjectListener(self._goListener)
        self.notifyListeners()

    def removeGraphicalObject(self, go: GraphicalObject):
        go.setSelected(False)
        go.removeGraphicalObjectListener(self._goListener)
        self._objects.remove(go)
        self.notifyListeners()

    #Vrati listu postojecih objekata.
    def getList(self):
        return self._roObjects

    #Prijava.
    def addDocumentModelListener(listener: DocumentModelListener):
        self._listeners.append(listener)

    #Odjava.
    def removeDocumentModelListener(listener: DocumentModelListener):
        self._listeners.remove(listener)

    #Obavještavanje.
    def notifyListeners(self):
        for documentModelListener in self._listeners:
            documentModelListener.documentChanged()
    
    #Vrati listu selektiranih objekata.
    def getSelectedObjects(self):
        return self._roSelectedObjects

    #Pomakni predani objekt u listi objekata na jedno mjesto kasnije...
    #Time će se on iscrtati kasniej (pa će time možda veći dio biti vidljiv)
    def increaseZ(self, go: GraphicalObject):
        index = self._objects.index(go)
        if index != -1 and index < len(self._objects) - 1:
            self._objects.insert(index, self._objects[index + 1])
            self._objects.insert(index + 1, go)

        self.notifyListeners()

    def decreaseZ(self, go: GraphicalObject):
        index = self._objects.index(go)
        if index != -1 and index > 0:
            self._objects.insert(index, self._objects[index + 1])
            self._objects.insert(index - 1, go)

        self.notifyListeners()

    #Pronađi postoji li u modelu neki objekt koji klik na točku koja je
    #predana kao argument selektira i vrati ga ili vrati null.
    #Točka selektira objekt kojemu je najbliža uz uvjet da ta udaljenost
    #nije veća od SELECTION_PROXIMITY. Status selektiranosti objekta ova
    #metoda NE dira.
    def findSelectedGraphicalObject(self, mousePoint: Point):
        possibleSelections = {}
        for go in self._objects:
            distance = go.selectionDistance(mousePoint)
            if distance <= SELECTION_PROXIMITY:
                possibleSelections[distance] = go
        
        if bool(possibleSelections):
            _min = SELECTION_PROXIMITY + 1
            for item in possibleSelections.items():
                if item[0] < _min:
                    _min = item[0]
            
            return possibleSelections[_min]
        
        return None

    #Pronađe li da u predanom objektu predana točka miša selektira neki hot-point.
    #Točka miša selektira onaj hot-point objekta kojemu je najbliža uz uvjet da ta
    #udaljenost nije veća od SELECTION_PROXIMITY. Vraća se index hot-pointa
    #kojeg bi predana točka selektirala ili -1 ako takve nema. Status selekcije
    #se pri tome NE dira.
    def findSelectedHotPoint(self, go: GraphicalObject, mousePoint: Point):
            for i in range(go.getNumberOfHotPoints()):
                if go.getHotPointDistance(i, mousePoint) <= SELECTION_PROXIMITY:
                    return i
            return -1
    

File: lab4/src/hr/fer/ooup/lab4/DocumentModel.java
package hr.fer.ooup.lab4;

import java.awt.Point;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

public class DocumentModel {
	
	public final static double SELECTION_PROXIMITY = 10;
	//Kolekcija svih grafièkih objekata:
	private List<GraphicalObject> objects = new ArrayList<>();
	//Read-Only proxy oko kolekcije grafièkih objekata:
	private List<GraphicalObject> roObjects = Collections.unmodifiableList(objects);
	//Kolekcija prijavljenih promatraèa:
	private List<DocumentModelListener> listeners = new ArrayList<>();
	//Kolekcija selektiranih objekata:
	private List<GraphicalObject> selectedObjects = new ArrayList<>();
	//Read-Only proxy oko kolekcije selektiranih objekata::
	private List<GraphicalObject> roSelectedObjects = Collections.unmodifiableList(selectedObjects);
	
	//Konstruktor...
	public DocumentModel() {
		objects = new ArrayList<GraphicalObject>();
		roObjects = Collections.unmodifiableList(objects);
		selectedObjects = new ArrayList<GraphicalObject>();
		roSelectedObjects = Collections.unmodifiableList(selectedObjects);
		listeners = new ArrayList<DocumentModelListener>();
	}
	
	//Brisanje svih objekata iz modela i potom
	//obavijeste svi promatraèi modela
	public void clear() {
		for (GraphicalObject go: objects) {
			go.removeGraphicalObjectListener(goListener);
		}
		objects.clear();
		selectedObjects.clear();
		notifyListeners();
	}
	
	//Dodavanje objekta u domkument
	public void addGraphicalObject(GraphicalObject go) {
		if(go.isSelected()) {
			selectedObjects.add(go);
		}
		objects.add(go);
		go.addGraphicalObjectListener(goListener);
		notifyListeners();
		
	}
	
	//Uklanjanje objekta iz dokumenta
	public void removeGraphicalObject(GraphicalObject go) {
		go.setSelected(false);
		go.removeGraphicalObjectListener(goListener);
		objects.remove(go);
		notifyListeners();
	}
	
	//Vrati listu postojeæih objekata
	public List<GraphicalObject> list() {
		return roObjects;
	}
	
	
	//Prijava...
	public void addDocumentModelListener(DocumentModelListener l) {
		listeners.add(l);
	}
	
	//Odjava...
	public void removeDocumentModelListener(DocumentModelListener l) {
		listeners.remove(l);
	}
	
	//Obavještavanje...
	public void notifyListeners() {
		for (DocumentModelListener documentModelListener : listeners) {
			documentModelListener.documentChange();
		}
	}
	
	
	//Promatraè koji æe biti registriran nad svim objektima crteža...
	private final GraphicalObjectListener goListener = new GraphicalObjectListener() {
		
		@Override
		public void graphicalObjectSelectionChanged(GraphicalObject go) {
			int index = selectedObjects.indexOf(go);
			if(go.isSelected()) {
				if (index == -1) {
					selectedObjects.add(go);
					
				} else {
					return;
				}
			}
			else{
					if (index != -1) {
						selectedObjects.remove(index);
					} else {
						return;
					}
			}
			notifyListeners();
			
			
			
		}
		
		@Override
		public void graphicalObjectChanged(GraphicalObject go) {
			notifyListeners();
			
		}
	};
	
	//Vrati listu selektiranih objekata
	public List<GraphicalObject> getSelectedObjects() {
		return roSelectedObjects;
	}
	
	//Pomakni predani objekt u listi objekata na jedno mjesto kasnije...
	public void increaseZ(GraphicalObject go) {
		int index = objects.indexOf(go);
		if (index != -1 && index > 0) {
			objects.set(index, objects.get(index - 1));
			objects.set(index - 1, go);
		}
		notifyListeners();
	}
	//Pomakni predani objekt u listi objekata ne jedno mjesto ranije...
	public void decreaseZ(GraphicalObject go) {
		int index = objects.indexOf(go);
		if (index != -1 && index > 0) {
			objects.set(index, objects.get(index - 1));
			objects.set(index - 1, go);
		}
		notifyListeners();
	}
 
	//Pronaði postoji li u modelu neki objekt koji klin na toèku koja je
	//predana kao argument selektira i vrati ga ili vrati null. 
	public GraphicalObject findSelectedGraphicalObject(Point mousePoint) {
		Map<Double, GraphicalObject> possibleSelections = new HashMap<Double, GraphicalObject>();
		for (GraphicalObject go : objects) {
			double distance = go.selectionDistance(mousePoint);
			if (distance <= SELECTION_PROXIMITY) {
				possibleSelections.put(distance, go);
			}
		}
		if (!possibleSelections.isEmpty()) {
			double min = SELECTION_PROXIMITY + 1;
			for (Entry<Double, GraphicalObject> entry : possibleSelections.entrySet()) {
				if (entry.getKey() < min) {
					min = entry.getKey();
				}
			}
			return possibleSelections.get(min);
		}
		
		return null;
	}
	
	//Pronaði da li u predanom objektu predana toèka miša selektira neki hot-point.
	//toèka miša selektira onaj hot-point objekata kojemu je najbliža uz uvjet da ta
	//udaljenost nije veæa od SELECTION_PROXIMITY. Vraæa se index hot-pointa kojeg bi
	//predana toèka selektirali ili -1 ako takve nema.
	public int findSelectedHotPoint(GraphicalObject go, Point mousePoint) {
		for (int i = 0; i < go.getNumberOfHotPoints(); i++) {
			if (go.getHotPointDistance(i, mousePoint) <= SELECTION_PROXIMITY) {
				return i;
			}
		}
		return -1;
		
	}
	
	public void deselectAll() {
		while(selectedObjects.size() > 0) {
			selectedObjects.get(0).setSelected(false);
		}
	}
	
	
	public List<GraphicalObject> findSelectedGraphicalObjects(Point mp) {
		List<GraphicalObject> Go = new ArrayList<GraphicalObject>();
		for (GraphicalObject go : objects) {
			if (go.selectionDistance(mp) <= SELECTION_PROXIMITY/2) {
				Go.add(go);
			}
		}
		return Go;
	}

	
}
